### PR TITLE
Spell out `Self` in async function return

### LINKED ISSUE #69276: Improve error message for "return type cannot contain a projection or `Self`..."
Example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6263bc1478126f230d8ef2dae9a5deb0)):

```
error: `async fn` return type cannot contain a projection or `Self` that references lifetimes from a parent scope
 --> src/lib.rs:4:33
  |
4 |     async fn new(i: &'a i32) -> Self {
  |                                 ^^^^
```

- Add a hint: try spelling out the type instead
- Add an error code with an entry in the error index explaining that this is a limitation of the compiler, tracked by #61949

### PR DESCRIPTION
**Pull request by csmoe**
Closes #69276
r? @tmandry

### CODE REVIEW

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -1632,47 +1632,56 @@ fn check_opaque_for_inheriting_lifetimes(tcx: TyCtxt<'tcx>, def_id: LocalDefId,
...
         opaque_identity_ty: Ty<'tcx>,
         generics: &'tcx ty::Generics,
+        ty: Option<Ty<'tcx>>,
     };
 
     impl<'tcx> ty::fold::TypeVisitor<'tcx> for ProhibitOpaqueVisitor<'tcx> {
         fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
             debug!("check_opaque_for_inheriting_lifetimes: (visit_ty) t={:?}", t);
+            self.ty = Some(t);
             if t == self.opaque_identity_ty { false } else { t.super_visit_with(self) }
```
Comment (tmandry) [inline src/librustc_typeck/check/mod.rs:1642]:
```suggestion
            if t != self.opaque_identity_ty && t.super_visit_with(self) {
                self.ty = Some(t);
                return true;
            }
            false
```

This avoids "resetting" `self.ty` in `visit_region`.

**Comment by tmandry** [comment]
Looks good, thanks!

@bors r+

**Comment by RalfJung** [comment]
Failed in https://github.com/rust-lang/rust/pull/72430#issuecomment-632280867
```
tidy check
tidy error: duplicate error code: 755
tidy error: /checkout/src/librustc_error_codes/error_codes.rs:440: E0755: include_str!("./error_codes/E0755.md"),
tidy error: /checkout/src/librustc_error_codes/error_codes.rs:620:     E0755, // `#[ffi_pure]` is only allowed on foreign functions
Checking which error codes lack tests...
some tidy checks failed
Found 496 error codes
Found 0 error codes with no tests
Done!
```
@bors r-

**Comment by csmoe** [comment]
r? @RalfJung 
error codes updated

**Comment by RalfJung** [comment]
Sorry, I can't review this, I don't know the code. I just saw a failure during a rollup.
r? @tmandry

**Comment by csmoe** [comment]
@tmandry ping

**Comment by csmoe** [comment]
This PR was approved, but dequeued from the merging list as error-code conflict.
r? @estebank

**Comment by estebank** [comment]
@csmoe there's another PR that has taken E0758 already (and I have one open for E0759). Update and I'll r+

**Comment by csmoe** [comment]
@estebank moved to E0760.

**Pull request merged**